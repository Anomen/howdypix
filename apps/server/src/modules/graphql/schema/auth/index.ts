export * from "./authTypes";
export * from "./authMutation";
