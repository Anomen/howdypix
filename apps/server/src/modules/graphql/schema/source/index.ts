export * from "./sourceTypes";
export * from "./sourceQuery";
