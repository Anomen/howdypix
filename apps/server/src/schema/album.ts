import { objectType } from "nexus";

export const Album = objectType({
  name: "Album",
  definition(t) {
    t.field("name", { type: "String" });
    t.field("dir", { type: "String" });
    t.field("source", { type: "String" });
  }
});
