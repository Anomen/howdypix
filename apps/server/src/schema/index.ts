export * from "./album";
export * from "./auth";
export * from "./user";
