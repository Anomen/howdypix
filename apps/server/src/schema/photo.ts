import { objectType } from "nexus";

export const Photo = objectType({
  name: "Photo",
  definition(t) {
    t.id("id");
    t.field("thumbnails", { type: "String", list: [false] });
  }
});
