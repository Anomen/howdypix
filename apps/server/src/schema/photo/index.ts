export * from "./photoTypes";
export * from "./photoQuery";
