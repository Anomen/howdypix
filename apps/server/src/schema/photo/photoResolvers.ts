import {
  NexusGenArgTypes,
  NexusGenFieldTypes,
  NexusGenRootTypes,
} from "@howdypix/graphql-schema/schema.d";
import { appDebug, generateThumbnailUrls } from "@howdypix/utils";
import { Photo as EntityPhoto } from "../../entity/Photo";
import { ApolloContext } from "../../types.d";
import { appConfig } from "../../config";
import { photoHelpers } from "../../helpers/photoHelpers";

const debug = appDebug("gql");

export const getPhotoResolver = () => async (
  root: {},
  args: NexusGenArgTypes["Query"]["getPhoto"],
  ctx: ApolloContext
): Promise<NexusGenFieldTypes["Query"]["getPhoto"]> => {
  debug(`Fetching photo ${args.file}.`);

  const photoRepository = ctx.connection.getRepository(EntityPhoto);
  const photo = await photoRepository.findOne({
    where: { dir: args.album, source: args.source, file: args.file },
  });

  if (photo) {
    const photoSteam = await photoHelpers.fetchPhotoSteam(
      ctx.connection,
      photo,
      args
    );
    const photoIndexInStream = photoSteam.findIndex(
      (sr) => sr.photoId === photo.id
    );

    return {
      id: photo.id.toString(),
      files: generateThumbnailUrls(appConfig.webapp.baseUrl, {
        file: photo.file,
        source: photo.source,
        name: photo.file,
        dir: photo.dir,
      }).map((tn) => tn.url),
      aperture: photo.processedAperture,
      shutter: photo.processedShutter,
      iso: photo.ISO,
      make: photo.make,
      model: photo.model,
      birthtime: photo.birthtime,
      next: photoSteam[photoIndexInStream + 1]?.photo.file ?? null,
      previous: photoSteam[photoIndexInStream - 1]?.photo.file ?? null,
      photoStream: photoSteam.map(
        (sr): NexusGenRootTypes["PhotoStreamThumbnail"] => ({
          id: String(sr.photo.id),
          file: String(sr.photo.file),
          thumbnails: generateThumbnailUrls(
            appConfig.webapp.baseUrl,
            sr.photo
          ).map((tn) => tn.url),
        })
      ),
    };
  }

  return null;
};
