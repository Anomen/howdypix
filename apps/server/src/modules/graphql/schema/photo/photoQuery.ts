import { queryField, stringArg } from "nexus";
import { getPhotoResolver } from "./photoResolvers";
import { EnhancedQuery } from "../../../../types";
import { withFilterByQueryArg, withOrderByQueryArg } from "../mixins";

export const getPhoto: EnhancedQuery = () =>
  queryField("getPhoto", {
    type: "PhotoDetail",
    args: {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      album: stringArg({ required: true }),
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      file: stringArg({ required: true }),
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      source: stringArg({ required: true }),
      ...withOrderByQueryArg(),
      ...withFilterByQueryArg(),
    },
    nullable: true,
    resolve: getPhotoResolver(),
  });
