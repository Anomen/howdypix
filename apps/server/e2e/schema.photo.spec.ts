import { NexusGenArgTypes } from "@howdypix/graphql-schema/schema.d";
import { getPhotoResolver } from "../src/schema/photo/photoResolvers";
import { initialize, resetValues, saveAllData } from "./schema.setup";

describe("photo resolver", () => {
  const { connection, reset } = initialize();

  beforeEach(async () => {
    jest.resetAllMocks();
    jest.restoreAllMocks();

    await reset();
  });

  describe("with a full database", () => {
    let { sources, albums, photos } = resetValues();

    beforeEach(() => {
      const newValues = resetValues();
      sources = newValues.sources;
      albums = newValues.albums;
      photos = newValues.photos;
    });

    ([
      [
        "should return photo without current filter / order",
        {
          album: photos[0].dir,
          source: photos[0].source,
          file: photos[0].file,
        },
      ],
      [
        "should return photo with current filter",
        {
          album: photos[0].dir,
          source: photos[0].source,
          file: photos[0].file,
          filterBy: { model: [photos[0].model] },
        },
      ],
      [
        "should return photo with current order",
        {
          album: photos[0].dir,
          source: photos[0].source,
          file: photos[0].file,
          orderBy: "DATE_ASC",
        },
      ],
      [
        "should return photo with both filter and order",
        {
          album: photos[0].dir,
          source: photos[0].source,
          file: photos[0].file,
          filterBy: { model: [photos[0].model] },
          orderBy: "DATE_ASC",
        },
      ],
    ] as [string, NexusGenArgTypes["Query"]["getPhoto"]][]).forEach(
      ([description, args]) => {
        it(description, async () => {
          // Initialize
          await saveAllData(connection, { sources, albums, photos });

          // Run
          const results = await getPhotoResolver()({}, args, {
            connection,
            user: null,
          });

          // Test
          expect(results).toMatchSnapshot();
        });
      }
    );
  });
});
