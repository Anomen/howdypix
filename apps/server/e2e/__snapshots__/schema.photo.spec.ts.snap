// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`photo resolver with a full database should return photo with both filter and order 1`] = `
Object {
  "aperture": null,
  "birthtime": 0,
  "files": Array [
    "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
  ],
  "id": "0",
  "iso": 22220,
  "make": "make1",
  "model": "model1",
  "next": "File2",
  "photoStream": Array [
    Object {
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
    Object {
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "file": "File3",
      "id": "3",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File3x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x2048.jpg",
      ],
    },
    Object {
      "file": "File5",
      "id": "5",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File5x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x2048.jpg",
      ],
    },
    Object {
      "file": "File6",
      "id": "6",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File6x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x2048.jpg",
      ],
    },
    Object {
      "file": "File8",
      "id": "8",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File8x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x2048.jpg",
      ],
    },
  ],
  "previous": null,
  "shutter": null,
}
`;

exports[`photo resolver with a full database should return photo with current filter 1`] = `
Object {
  "aperture": null,
  "birthtime": 0,
  "files": Array [
    "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
  ],
  "id": "0",
  "iso": 22220,
  "make": "make1",
  "model": "model1",
  "next": null,
  "photoStream": Array [
    Object {
      "file": "File8",
      "id": "8",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File8x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x2048.jpg",
      ],
    },
    Object {
      "file": "File6",
      "id": "6",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File6x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x2048.jpg",
      ],
    },
    Object {
      "file": "File5",
      "id": "5",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File5x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x2048.jpg",
      ],
    },
    Object {
      "file": "File3",
      "id": "3",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File3x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x2048.jpg",
      ],
    },
    Object {
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
  ],
  "previous": "File2",
  "shutter": null,
}
`;

exports[`photo resolver with a full database should return photo with current order 1`] = `
Object {
  "aperture": null,
  "birthtime": 0,
  "files": Array [
    "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
  ],
  "id": "0",
  "iso": 22220,
  "make": "make1",
  "model": "model1",
  "next": "File1",
  "photoStream": Array [
    Object {
      "file": "File0",
      "id": "0",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
      ],
    },
    Object {
      "file": "File1",
      "id": "1",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File1x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File1x2048.jpg",
      ],
    },
    Object {
      "file": "File2",
      "id": "2",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File2x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File2x2048.jpg",
      ],
    },
    Object {
      "file": "File3",
      "id": "3",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File3x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x2048.jpg",
      ],
    },
    Object {
      "file": "File4",
      "id": "4",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File4x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File4x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File4x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File4x2048.jpg",
      ],
    },
    Object {
      "file": "File5",
      "id": "5",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File5x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x2048.jpg",
      ],
    },
  ],
  "previous": null,
  "shutter": null,
}
`;

exports[`photo resolver with a full database should return photo without current filter / order 1`] = `
Object {
  "aperture": null,
  "birthtime": 0,
  "files": Array [
    "http://localhost:3000/files/@source0:albumDir0/File0x200.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x600.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x1280.jpg",
    "http://localhost:3000/files/@source0:albumDir0/File0x2048.jpg",
  ],
  "id": "0",
  "iso": 22220,
  "make": "make1",
  "model": "model1",
  "next": "File8",
  "photoStream": Array [
    Object {
      "file": "File8",
      "id": "8",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File8x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File8x2048.jpg",
      ],
    },
    Object {
      "file": "File7",
      "id": "7",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File7x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File7x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File7x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File7x2048.jpg",
      ],
    },
    Object {
      "file": "File6",
      "id": "6",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File6x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File6x2048.jpg",
      ],
    },
    Object {
      "file": "File5",
      "id": "5",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File5x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File5x2048.jpg",
      ],
    },
    Object {
      "file": "File4",
      "id": "4",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File4x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File4x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File4x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File4x2048.jpg",
      ],
    },
    Object {
      "file": "File3",
      "id": "3",
      "thumbnails": Array [
        "http://localhost:3000/files/@source0:albumDir0/File3x200.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x600.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x1280.jpg",
        "http://localhost:3000/files/@source0:albumDir0/File3x2048.jpg",
      ],
    },
  ],
  "previous": null,
  "shutter": null,
}
`;
